### PR TITLE
Convert highlight helper to Typescript

### PR DESCRIPTION
**Pull request by daniellesniak**
See flarum/framework#3533

### CODE REVIEW

**Comment by dsevillamartin** [comment]
Should the types be removed from the jsdoc comment? 🤔

I feel like there's no reason to have them there twice. The JSDoc is still helpful for the description of the arguments.

**Comment by daniellesniak** [comment]
You're right, I fixed that.

File: js/src/common/helpers/highlight.ts [hunk 1]
```diff
@@ -4,21 +4,20 @@ import { truncate } from '../utils/string';
...
- * @param {String|RegExp} phrase The word or words to highlight.
- * @param {Integer} [length] The number of characters to truncate the string to.
+ * @param string The string to highlight.
+ * @param phrase The word or words to highlight.
+ * @param [length] The number of characters to truncate the string to.
  *     The string will be truncated surrounding the first match.
- * @return {Object}
  */
-export default function highlight(string, phrase, length) {
+export default function highlight(string: string, phrase: string | RegExp, length: number): Object | string {
```
Comment (w-4) [inline js/src/common/helpers/highlight.ts:12]: The return type should be set to `Mithril.Vnode` here

File: js/src/common/helpers/highlight.ts [hunk 2]
```diff
@@ -4,21 +4,20 @@ import { truncate } from '../utils/string';
...
   if (!phrase && !length) return string;
 
   // Convert the word phrase into a global regular expression (if it isn't
   // already) so we can search the string for matched.
-  const regexp = phrase instanceof RegExp ? phrase : new RegExp(phrase, 'gi');
+  const regexp: RegExp = phrase instanceof RegExp ? phrase : new RegExp(phrase, 'gi');
```
Comment (w-4) [inline js/src/common/helpers/highlight.ts:17]: Personal preference, I'd remove the type here and the two below, since they are automatically inferred

**Comment by w-4** [comment]
also: `.tsx` file extension

File: js/src/common/helpers/highlight.ts (hunk 1 above)
Comment (daniellesniak) [reply js/src/common/helpers/highlight.ts]: fixed

File: js/src/common/helpers/highlight.ts (hunk 2 above)
Comment (daniellesniak) [reply js/src/common/helpers/highlight.ts]: Sounds good to me, done

File: js/src/common/helpers/highlight.tsx [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
- * @param {String|RegExp} phrase The word or words to highlight.
- * @param {Integer} [length] The number of characters to truncate the string to.
+ * @param string The string to highlight.
+ * @param phrase The word or words to highlight.
+ * @param [length] The number of characters to truncate the string to.
  *     The string will be truncated surrounding the first match.
- * @return {Object}
  */
-export default function highlight(string, phrase, length) {
+export default function highlight(string: string, phrase: string | RegExp, length: number): Mithril.Vnode | string {
```
Comment (datitisev) [inline js/src/common/helpers/highlight.tsx:13]: Length should be `length?: number` as it is an optional parameter.

Comment (daniellesniak) [reply js/src/common/helpers/highlight.tsx]: In typescript all parameters are required by default so yes - it should be, fixed.

Comment (datitisev) [reply js/src/common/helpers/highlight.tsx]: True, but we don't have TS errors or warnings enabled. The TS is just converted into plain JS - type hinting is mainly to help during development 🙂

**Comment by daniellesniak** [comment]
I made length parameter optional, everything else look fine to me.

**Review by askvortsov1 — APPROVED** [review]
Makes sense to me, thanks!

**Comment by datitisev** [comment]
I wonder - does `m.trust()` only return a `Mithril.Vnode`? I am unable to check right now.

**Comment by daniellesniak** [comment]
It's written in documentation that m.trust() returns Vnode. The same return type is in interface.

Docs:
trust

Interface:
hyperscript interface

**Comment by datitisev** [comment]
> Should I change return type to Mithril.Vnode<any, any> as in interface instead of Mithril.Vnode?

I think so, yeah.

**Review by dsevillamartin — APPROVED** [review]
LGTM!

**Pull request merged**